modules/OS/module/src/os_sem_posix.c

```diff
@@ -94,10 +139,60 @@ timespec_init_timeout__(struct timespec* ts, uint64_t us)
     ts->tv_nsec %= 1000000000;
 }
 
-int
-os_sem_take_timeout(os_sem_t sem, uint64_t usecs)
+static int
+os_sem_take_timeout_efd__(os_sem_t sem, uint64_t usecs)
 {
-    VALIDATE(sem);
+    AIM_TRUE_OR_DIE(sem->efd > 0);
+
+
+    if(usecs == 0) {
+        /** Normal wait. */
+        return os_sem_take(sem);
+    }
+    else {
+
+        /** poll() with timeout (in ms) */
+        struct pollfd fds;
+
+        fds.fd = sem->efd;
+        fds.events = POLLIN;
+        fds.revents = 0;
+
+        int to = usecs / 1000;
+        int rv;
+
+        switch( (rv = poll(&fds, 1, to)) )
+            {
+            case 0:
+                {
+                    /* Timed out */
+                    return -1;
+                }
+                break;
+            case 1:
+                {
+                    if(fds.revents & POLLIN) {
+                        /* Descriptor Ready */
+                        return os_sem_take(sem);
```

Comment:
This works for a single consumer, but if there are multiple consumers then all of them would wake up and only one would successfully take the semaphore. The rest would be stuck in read() until another os_sem_give.

You could do the read nonblocking and go back to the poll loop if it fails.